internal/mongodoc/resource_test.go

```diff
@@ -0,0 +1,369 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package mongodoc_test
+
+import (
+	"time"
+
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+	"gopkg.in/mgo.v2/bson"
+
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+)
+
+const fingerprint = "0123456789abcdef0123456789abcdef0123456789abcdef"
+
+var (
+	_ = gc.Suite(&ResourceSuite{})
+	_ = gc.Suite(&ResourcesSuite{})
+)
+
+type ResourceSuite struct{}
+
+func (s *ResourceSuite) TestNewResourceQueryWithRevision(c *gc.C) {
+	cURL := charm.MustParseURL("cs:trusty/spam-2")
+
+	query := mongodoc.NewResourceQuery(cURL, "eggs", 3)
+
+	c.Check(query, jc.DeepEquals, bson.D{
+		{"unresolved-charm-url", charm.MustParseURL("cs:spam")},
+		{"name", "eggs"},
+		{"revision", 3},
+	})
+}
+
+func (s *ResourceSuite) TestNewResourceQueryWithoutRevision(c *gc.C) {
+	cURL := charm.MustParseURL("cs:trusty/spam-2")
+
+	query := mongodoc.NewResourceQuery(cURL, "eggs", -1)
+
+	c.Check(query, jc.DeepEquals, bson.D{
+		{"unresolved-charm-url", charm.MustParseURL("cs:spam")},
+		{"name", "eggs"},
+	})
+}
+
+func (s *ResourceSuite) TestValidateFull(c *gc.C) {
+	doc := mongodoc.Resource{
```

Comment:
Let's make all these into a table-driven test for consistency with all the other code, please. It makes it much easier to see the consistency between the tests.